packages/credentials/src/ctype/CTypeLoader.ts

```diff
@@ -32,9 +37,69 @@ export function newCachingCTypeLoader(
   })
 
   async function getCType(id: ICType['$id']): Promise<ICType> {
-    const ctype: ICType = ctypes.get(id) ?? (await loadCType(id))
+    let ctype = ctypes.get(id)
+    if (ctype) {
+      return ctype
+    }
+    ctype = await cTypeLoader(id)
+    verifyDataStructure(ctype)
+    if (id !== ctype.$id) {
+      throw new SDKErrors.CTypeIdMismatchError(ctype.$id, id)
+    }
     ctypes.set(ctype.$id, ctype)
     return ctype
   }
   return getCType
 }
+
+/**
+ * Recursively traverses a (nested) CType's definition to load definitions of CTypes referenced within.
+ *
+ * @param cType A (nested) CType containg references to other CTypes.
+ * @param cTypeLoader A function with which to load CType definitions.
+ * @returns An array of CType definitions which were referenced in the original CType or in any of its composite CTypes.
+ */
+export async function loadNestedCTypeDefinitions(
+  cType: ICType,
+  cTypeLoader: CTypeLoader
+): Promise<ICType[]> {
+  const fetchedCTypeIds = new Set<string>()
+  const fetchedCTypeDefinitions: ICType[] = []
+
+  // Don't fetch the original CType
+  fetchedCTypeIds.add(cType.$id)
+
+  async function extractRefsFrom(value: unknown): Promise<void> {
+    if (typeof value !== 'object' || value === null) {
+      return
+    }
+
+    if ('$ref' in value) {
+      const ref = (value as { $ref: unknown }).$ref
+      if (typeof ref === 'string' && ref.startsWith('kilt:ctype:')) {
+        const cTypeId = ref.split('#/')[0] as ICType['$id']
+
+        if (!fetchedCTypeIds.has(cTypeId)) {
+          fetchedCTypeIds.add(cTypeId)
+          const referencedCType = await cTypeLoader(cTypeId)
+
+          if (isICType(referencedCType)) {
+            fetchedCTypeDefinitions.push(referencedCType)
+          } else {
+            throw new Error(`Failed to load referenced CType: ${cTypeId}`)
+          }
+
+          await extractRefsFrom(referencedCType.properties)
+        }
+      }
+      return
+    }
+
+    // Process all values in the object. Also works for arrays
+    await Promise.all(Object.values(value).map(extractRefsFrom))
+  }
+
+  await extractRefsFrom(cType.properties)
+
+  return fetchedCTypeDefinitions
+}
```

Comment:
What if we have self-referential? or a CType that references another one that references the first? 

We would be stuck in a loop. Can we prevent this by checking if a CType `$id` exists already? hmm, I guess this doesn't work.

What if we have a max depth? I mean, shouldn't we already have restrictions in the Meta-Schema or on-chain. Therefore, we should use this otherwise, we can get an infinite loop.

Reply:
> What if we have self-referential? or a CType that references another one that references the first?
> 
> We would be stuck in a loop. Can we prevent this by checking if a CType `$id` exists already? hmm, I guess this doesn't work.

It does work, and we do it on line 82. So no getting stuck in a loop.

Reply:
Ok, I did a lot of digging to see how it all works. I understand now. My brain hurts.